File: test_arx_sysid.m
% Test ARX model identification - Example from documentation
% This code is based on the example at
% https://www.mathworks.com/help/ident/ref/arx.html

clear all
rng(0)

% Specify a polynomial model sys0 with the ARX structure.
A = [1  -1.5  0.7];
B = [0 1 0.5];
sys0 = idpoly(A,B);

% Generate a measured input signal u that contains random 
% binary noise and an error signal e
N = 300;
u = iddata([],idinput(N,'rbs'));
u.InputName = 'u';
e = iddata([],randn(N,1));
e.InputName = 'e';
y = sim(sys0,[u e]);
y.OutputName = 'y';

% Combine y and u into a single iddata object z.
z = [y,u];

t = z.Ts.*(z.Tstart:N)';

% Save data
data_est = table( ...
    t, u.InputData, e.InputData, y.OutputData, ...
    'VariableNames', {'t', 'u', 'e', 'y'} ...
);
writetable(data_est, "test_arx_sysid_data.csv")

% Estimate a new ARX model using z and the same 
% polynomial orders and input delay as the 
% original model.
sys = arx(z,[2 2 1]);

% Display results
present(sys)
disp("getpvec(sys):")
fprintf("[%s]\n", strjoin(compose("%.6f ", getpvec(sys))))

assert(isequal( ...
    round(getpvec(sys), 6), ...
    [-1.523848  0.713374  0.999994  0.474829 ]' ...
))

% Simulate model to produce a prediction
e0 = iddata([],zeros(N, 1));
y_pred = sim(sys,[u e0]);
y_pred.OutputName = 'y_pred';

% Calculate root-mean-squared error
rmse = sqrt(mean((y.OutputData - y_pred.OutputData).^2));
fprintf("Root-mean-squared-error: %.3f\n", rmse)

figure(1); clf
subplot(211)
plot(u)
legend()

subplot(212)
plot(y); hold on
plot(y_pred)
grid on
legend()

% OUTPUT:
%
% sys =                                                                         
% Discrete-time ARX model: A(z)y(t) = B(z)u(t) + e(t)                           
%   A(z) = 1 - 1.524 (+/- 0.02767) z^-1 + 0.7134 (+/- 0.02717) z^-2             
%                                                                               
%   B(z) = 1 (+/- 0.05831) z^-1 + 0.4748 (+/- 0.06362) z^-2                     
%                                                                               
% Sample time: 1 seconds                                                        
%                                                                               
% Parameterization:                                                             
%    Polynomial orders:   na=2   nb=2   nk=1                                    
%    Number of free coefficients: 4                                             
%    Use "polydata", "getpvec", "getcov" for parameters and their uncertainties.
%                                                                               
% Status:                                                                       
% Estimated using ARX on time domain data "z".                                  
% Fit to estimation data: 81.36% (prediction focus)                             
% FPE: 1.025, MSE: 0.9846                                                       
% More information in model's "Report" property.                                
% getpvec(sys):
% [-1.523848  0.713374  0.999994  0.474829 ]   

File: test_arx_sysid.py
# Test ARX model identification and compare to example from MATLAB
# documentation.
#
# This code is based on the example at
# https://apmonitor.com/wiki/index.php/Apps/ARXTimeSeries

from gekko import GEKKO
import numpy as np
import pandas as pd
import matplotlib.pyplot as plt

filename = "test_arx_sysid_data.csv"
data_est = pd.read_csv(filename)

# load data and parse into columns
t = data_est.t  
u = data_est.u
y = data_est.y

m = GEKKO()

# system identification
na = 2 # output coefficients
nb = 2 # input coefficients
nk = 1
y_pred, p, K = m.sysid(t, u, y, na, nb, nk, pred='mean', shift='calc')

# Display results
print(p)
print(f"K: {K}")

# Calculate root-mean-squared error
rmse = np.sqrt(np.mean((y.values.reshape(-1, 1) - y_pred) ** 2))
print(f"Root-mean-squared-error: {rmse:.3f}")

plt.figure(figsize=(9, 5.5))
plt.subplot(2,1,1)
plt.plot(t, u, label='u')
plt.legend()
plt.grid()

plt.subplot(2,1,2)
plt.plot(t, y)
plt.plot(t, y_pred)
plt.legend(['y', 'y_pred'])
plt.xlabel('Time')
plt.grid()

plt.tight_layout()
plt.savefig("test_arx_sysid_py_plot.pdf")
plt.show()
